### DISCUSSION TITLE
verification code in web apps?

### DISCUSSION THREAD

**Discussion by ghost** [framesurge/perseus#211] [Q&A]
How to generate verification code to prevent DDoS, or just use Google CAPTCHA?

**Comment by arctic-hen7** [comment]
Depends on your use-case. If you want to protect an entire site, maybe look into Cloudflare's offerings, I believe they have some stuff for this. If it's a login form or the like, you can just follow the usual Captcha tutorials, substituting JS functions for their `web_sys` equivalents (`web_sys` wraps most things you need). Alternatively, you could take a look at `examples/core/js_interop` in this repo for a basic reference on how to integrate some JS into your app. This would let you use NPM libraries and the like if you have requirements that are solved by a JS package.

Could you perhaps clarify exactly what you want to do so I can give a more specific, and hopefully more helpful, answer?

**Reply by ghost** [reply]
I want to use pure Rust to generate CAPTCHA, without any JavaScript or HTML.

**Reply by lukechu10** [reply]
If you want to use Google captcha, you still need to include the google captcha js. You can then use `wasm-bindgen` to call the js code from your rust code.

**Reply by arctic-hen7** [reply]
As Luke has said, this won't be possible unfortunately with pure Rust. The state of the Web right now is that most popular libraries for things like this are written in JS, and only some have pure Rust ports. The best thing to do is to call the JS from your Rust code.

However, as the Wasm ecosystem grows, I have no doubt that, eventually, these kinds of things will be implemented without JS! It'll just take some time.

**Reply by ghost** [reply]
> As Luke has said, this won't be possible unfortunately with pure Rust.

Rust is a turing-complete language, which means it can do everything.

**Reply by lukechu10** [reply]
> > As Luke has said, this won't be possible unfortunately with pure Rust.
> 
> Rust is a turing-complete language, which means it can do everything.

That doesn't really mean anything. Plenty of things are "turing-complete", including powerpoint and minecraft.

**Reply by arctic-hen7** [reply]
@lojowi thus is true technically, but you would have to implement every little detail yourself. When I say "not possible", I suppose I really mean "infeasibly inconvenient unless you're willing to devotr a lot of time to it". Now, if you wanted to make a Rust port of Captcha (I'm not aware of any right now), that would be extremely helpful to the community!

Until either you or someone else does that though, this would be very time-consuming to do in pure Rust, without calling out to a JS library (which, again, would be the simplest thing).
